jax/_src/lax/control_flow/for_loop.py

```diff
@@ -372,3 +386,114 @@ def write(v: core.Var, val: Any) -> None:
+
+for_p = core.Primitive('for')
+for_p.multiple_results = True
+
+### Tracing utilities
+
+def _hoist_consts_to_refs(jaxpr: core.Jaxpr) -> core.Jaxpr:
+  num_consts = len(jaxpr.constvars)
+
+  # Note that this function is meant for use w/ `for_loop` since it assumes
+  # that the index is the first argument and preserves this after hoisting
+  # consts.
+  def _hoist(i, *consts_args):
+    const_refs, args = split_list(consts_args, [num_consts])
+    # We immediately read the const values out of the `Ref`s.
+    consts = [r[()] for r in const_refs]
+    return core.eval_jaxpr(jaxpr, consts, i, *args)
+  const_avals = [ShapedArrayRef(var.aval.shape, var.aval.dtype) for var in
+                 jaxpr.constvars]
+  i_aval, *arg_avals = [var.aval for var in jaxpr.invars]
+  in_avals = [i_aval, *const_avals, *arg_avals]
+  hoisted_jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(
+      lu.wrap_init(_hoist), in_avals)
+  assert not consts, "All consts should have been converted to refs"
+  return hoisted_jaxpr
+
+def _trace_to_jaxpr_with_refs(f, state_tree: PyTreeDef,
+                              state_avals: Sequence[core.AbstractValue]
+                              ) -> Tuple[core.Jaxpr, List[Any], PyTreeDef]:
+  f, out_tree_thunk = flatten_fun_nokwargs(
+      lu.wrap_init(f), treedef_tuple((tree_structure(0), state_tree)))
+  jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(
+      f, state_avals)
+  return jaxpr, consts, out_tree_thunk()
+
+def val_to_ref_aval(x) -> ShapedArrayRef:
+  aval = core.raise_to_shaped(core.get_aval(x))
+  if type(aval) is not core.ShapedArray:
+    raise Exception(f"can't make ref from {x}")
+  return ShapedArrayRef(aval.shape, aval.dtype)
+
+def for_loop(nsteps: int, body: Callable[[Array, Ref[S]], None], init_state: S) -> S:
+  """A for-loop combinator that allows read/write semantics in the loop body.
+
+  `for_loop` is a higher-order function that enables writing loops that can be
+  staged out in JIT-ted JAX computations. Unlike `jax.lax.fori_loop`, it allows
+  mutation in its body using `Ref`s.
+
+  `for_loop` will initialize `Ref`s with the values in `init_state`. Each
+  iteration, `body` will be called with the current `Ref`s, which can be read
+  from and written to using `ref_get` and `ref_set`.
+
+  `for_loop` is semantically equivalent to the following Python code:
+
+  ```python
+  def for_loop(nsteps, body, init_state):
+    refs = tree_map(make_ref, init_state)
+    for i in range(nsteps):
+      body(i, refs)
+    return tree_map(ref_get, refs)
+  ```
+
+  Args:
+    nsteps: Number of iterations
+    body: A callable that takes in the iteration number as its first argument
+          and `Ref`s corresponding to `init_state` as its second argument.
+          `body` is free to read from and write to its `Ref`s. `body` should
+          not return anything.
+    init_state: A Pytree of JAX-compatible values used to initialize the `Ref`s
+                that will be passed into the for loop body.
```

Comment:
```suggestion
    body: A callable that takes in the iteration number as its first argument
      and `Ref`s corresponding to `init_state` as its second argument.
      `body` is free to read from and write to its `Ref`s. `body` should
       not return anything.
    init_state: A Pytree of JAX-compatible values used to initialize the `Ref`s
      that will be passed into the for loop body.
```

I think something like that is how we usually indent these docstrings.